ipol_demo/modules/demorunner/demorunner.py

```diff
@@ -28,847 +28,713 @@
+    virtualenv.cli_run([venv_path])
+
+    cmd = [pip_bin, "install", "-r", os.path.join(src_dir, packages_file)]
+    cmd = shlex.split(" ".join(cmd))
+    install_proc = Popen(cmd, stdout=PIPE, stderr=PIPE)
+    _, stderr_data = install_proc.communicate()
+
+    if install_proc.returncode != 0:
+        raise IPOLCompileError(
+            f"Construct failed. Can't create virtualenv:\n{stderr_data}"
+        )
+
+
+# ---------------------------------------------------------------------------
+# Algorithm runner
+# ---------------------------------------------------------------------------
+def run_algo(
+    demo_id: int,
+    work_dir: str,
+    bin_path: str,
+    ddl_run: str,
+    params: dict,
+    res_data: dict,
+    timeout: int,
+) -> int:
+    """
+    the core algo runner
+    """
+    lock_path = os.path.join(bin_path, settings.compilation_lock_filename)
+    while construct_is_locked(lock_path):
+        time.sleep(1)
+    rd = run_demo_base.RunDemoBase(bin_path, work_dir, logger, timeout)
+    rd.set_algo_params(params)
+    rd.set_algo_info(res_data["algo_info"])
+    rd.set_MATLAB_path(settings.MATLAB_path)
+    rd.set_extra_path(settings.extra_path)
+    rd.set_demo_id(demo_id)
+    rd.set_commands(ddl_run)
+
+    rd.set_share_demoExtras_dirs(settings.share_demoExtras_dir, demo_id)
+
+    if not isinstance(ddl_run, str):
+        return -1  # Bad run syntax: not a string
+
+    # Substitute variables and run algorithm
+    cmd = variable_substitution(ddl_run, demo_id, params)
+
+    rd.run_algorithm(cmd, lock_run)
+
+    res_data["params"] = rd.get_algo_params()  # Should not be used
+    # Info interface --> algo
+    res_data["algo_info"] = rd.get_algo_info()
+    return 0
+
+
+def variable_substitution(ddl_run: str, demo_id: int, params: dict) -> str:
+    """
+    Replace the variables with its values and return the command to be executed
+    """
+    params["demoextras"] = os.path.join(settings.share_demoExtras_dir, str(demo_id))
+    params["matlab_path"] = settings.MATLAB_path
+    params["bin"] = get_bin_dir(demo_id)
+    params["virtualenv"] = get_bin_dir(demo_id) + "venv"
+    return Template(ddl_run).substitute(**params)
+
+
+def get_bin_dir(demo_id: int) -> str:
+    """
+    Returns the directory with the peer-reviewed author programs
+    """
+    return os.path.join(settings.main_bin_dir, f"{demo_id}bin/")
+
+
+def read_workdir_file(work_dir: str, filename: str) -> str:
+    """
+    Read a text file from the working directory and return its contents as UTF-8
+    """
+    full_file = os.path.join(work_dir, filename)
+    content = ""
+    if os.path.isfile(full_file):
+        with codecs.open(full_file, "r", "utf8", errors="replace") as f:
+            content = f.read()
+    return content
+
+
+@app.post("/exec_and_wait/{demo_id}", status_code=200)
+async def exec_and_wait(
+    demo_id: int,
+    key: str,
+    ddl_run: str,
+    parameters: Annotated[str, Body()],
```

Comment:
can we replace `str` with a dictionary?

Reply:
I haven't found the way to make it work like that